tests/test_cli_compile.py

```diff
@@ -2190,6 +2190,49 @@ def test_multiple_extras(fake_dists, runner, make_module, fname, content, extra_
     assert "extra ==" not in out.stderr
 
 
+@pytest.mark.network
+@pytest.mark.parametrize(("fname", "content"), METADATA_TEST_CASES)
+def test_build_system_requires(
+    request, fake_dists, runner, make_module, fname, content
+):
+    """
+    Test that requirements needed to build a package are included
+    """
+    meta_path = make_module(fname=fname, content=content)
+    out = runner.invoke(
+        cli,
+        [
+            "-n",
+            "--find-links",
+            fake_dists,
+            "--allow-unsafe",
+            "--no-annotate",
+            "--no-emit-options",
+            "--no-header",
+            "--build-system-requires",
```

Comment:
We also need a test case showing how to only extract the build system requirements without the runtime requirements.
Oh, and how about being able to separately request sdist build requirements, wheel build requirements and editable install requirements?

Reply:
> We also need a test case showing how to only extract the build system requirements without the runtime requirements.

Makes sense given that the CLI flag presents this an option. Will fix.

> Oh, and how about being able to separately request sdist build requirements, wheel build requirements and editable install requirements

Good catch, I had missed that build systems can state additional requirements for sdist and wheel. But I cannot find any information on editable installs, do you remember where you learned about it?

I am however unsure about the use case for locking sdist and wheel requirements separately. What do you have in mind?
If there is no clear use case, is it worth increasing the size of the UI and the test matrix?

Reply:
> But I cannot find any information on editable installs, do you remember where you learned about it?

https://peps.python.org/pep-0660/#get-requires-for-build-editable

> I am however unsure about the use case for locking sdist and wheel requirements separately. What do you have in mind?

The fact that they are dynamically calculated suggests that they may be different or even conflicting.

Also, building a wheel from a Git checkout may be different than building a wheel from an sdist due to possible inconsistencies in what files are bundles in a wheel.
I imagine the same may happen with the editable checkout. Also, I could see a case when the editable install deps include some development helpers that further restrict the constraints.

> If there is no clear use case, is it worth increasing the size of the UI and the test matrix?

It's probably worth thinking about edge cases. One case that I just realized is that the build deps can have env markers (both direct and transitive). This means that there could be different constraints produced depending on which OS/arch/Python version is used to generate them. This is an interesting case to explore. I currently experiment with having a matrix of constraint files per env combo, for runtime. Having that for build deps will probably be useful for non pure Python projects...

Reply:
>> We also need a test case showing how to only extract the build system requirements without the runtime requirements.
> 
> Makes sense given that the CLI flag presents this as an option. Will fix.

Now that I read the code again I am not sure what I was referring to. Instead it seems to me that we would have to add another flag to accomplish this e.g. `--no-install-requires`. This opens up questions like how such an option should interact with the `--extra` argument and I am starting to think that this should be treated as a separate issue.

I am also thinking about what would be a good interface for telling `pip-compile` what build requirements to include. I'm leaning towards replacing the previously proposed `--build-system-requires` flag with an `--all-build` flag and a `--build` parameter that works similarly to the `--all-extra` flag and `--extra` parameter respectively but takes `editable`, `sdist` or `wheel` as arguments.

Whatever the shape of the interface I think I need to spend some time exploring how the hooks work and how they can be tested. I will try to make some time for this over the coming weeks.

Reply:
Extras are tightly coupled with the normal runtime deps. So if they aren't used, they shouldn't be pulled in...

Reply:
I have implemented support for this but I am not sure I like it because it adds complexity to both the public interface and the implementation and it is not clear to me that anyone will actually use the feature.